### PR TITLE
Add New-GitHubPullRequest command

### PR DESCRIPTION
**Pull request by rjmholt**
Resolves https://github.com/microsoft/PowerShellForGitHub/issues/30.

Adds a command for creating new pull requests.

Outstanding questions:

- Are the parameters ok, should they be changed?
- Are more examples needed?
- How do test?

### CODE REVIEW

File: GitHubPullRequests.ps1 [hunk 1]
```diff
@@ -163,3 +163,192 @@ function Get-GitHubPullRequest
...
+        [switch] $Draft,
+
+        [string] $AccessToken,
+
+        [switch] $NoStatus
+    )
+
+    Write-InvocationLog
+
+    if ($HeadOwner)
```
[the remark this answers is not in the document]
Comment (rjmholt) [reply GitHubPullRequests.ps1]: I believe PowerShell's conversion [does the same thing](https://github.com/PowerShell/PowerShell/blob/dbbb5110bf5b21c04929ac0e3df3dae86037b220/src/System.Management.Automation/engine/LanguagePrimitives.cs#L988-L1004). There's no strange number coercion like in JavaScript.

I can change it anyway, but given that a variable is guaranteed to be a string I think they're equivalent

File: GitHubPullRequests.ps1 [hunk 2]
```diff
@@ -163,3 +163,192 @@ function Get-GitHubPullRequest
...
+    {
+        $postBody['draft'] = $true
+        $acceptHeader = 'application/vnd.github.shadow-cat-preview+json'
+    }
+
+    $restParams = @{
+        'UriFragment' = $uriFragment
+        'Method' = 'Post'
+        'Description' = $description
+        'Body' = ConvertTo-Json -InputObject $postBody -Compress
```
[the remark this answers is not in the document]
Comment (rjmholt) [reply GitHubPullRequests.ps1]: Not a good if you're logging, but saves some memory, IO, etc

File: GitHubPullRequests.ps1 [hunk 3]
```diff
@@ -163,3 +163,192 @@ function Get-GitHubPullRequest
...
+    .PARAMETER RepositoryName
+        Name of the repository.
+        If not supplied here, the DefaultRepositoryName configuration property value will be used.
+    
+    .PARAMETER Uri
+        Uri for the repository.
+        The OwnerName and RepositoryName will be extracted from here instead of needing to provide
+        them individually.
+    
+    .PARAMETER Title
```
[the remark this answers is not in the document]
Comment (rjmholt) [reply GitHubPullRequests.ps1]: Happy to!

File: GitHubPullRequests.ps1 (hunk 1 above)
[the remark this answers is not in the document]
Comment (HowardWolosky) [reply GitHubPullRequests.ps1]: Based on [this other section](https://github.com/PowerShell/PowerShell/blob/dbbb5110bf5b21c04929ac0e3df3dae86037b220/src/System.Management.Automation/engine/LanguagePrimitives.cs#L1025-L1028), they're not equivalent, as PowerShell only considers _empty_ strings to be `$null`/`$false`.  In this case, this is _additionally_ saying that a string of pure whitespace should also be considered `$false` as well.

> Aside: TIL that you can have multiple lines selected when referencing a section of a file on GitHub.  Neat.

[the remark this answers is not in the document]
Comment (rjmholt) [reply GitHubPullRequests.ps1]: Oh I read it as `NullOrEmpty`!

File: GitHubPullRequests.ps1 [hunk 4]
```diff
@@ -163,3 +163,192 @@ function Get-GitHubPullRequest
...
+    $uriFragment = "/repos/$OwnerName/$RepositoryName/pulls"
+    $description = "Creating pull request $Title for $RepositoryName"
+
+    $postBody = @{
+        'title' = $Title
+        'head' = $Head
+        'base' = $Base
+    }
+
+    if ($Body)
```
[the remark this answers is not in the document]
Comment (rjmholt) [reply GitHubPullRequests.ps1]: I just thought about this again. It's possible to open a PR that has only whitespace in the body I think.

The current implementation doesn't put a restriction on `$Title`.

So the possible approaches are:
- Allow whitespace as `$Title` or `$Body`
- Silently drop `$Body` when it's whitespace and throw when `$Title` is whitespace
- Throw when `$Body` or `$Title` is whitespacet

**Comment by HowardWolosky** [comment]
Hey @rjmholt  -- just checking back in with you after the long weekend.  I'm just looking to see if you were intending to iterate any further on this, or if I should make the [updates to this](https://github.com/microsoft/PowerShellForGitHub/pull/111#discussion-diff-287150230R120) myself after merging in your change.

Thanks again for this contribution!

**Comment by rjmholt** [comment]
> Hey @rjmholt Robert Holt FTE -- just checking back in with you after the long weekend. I'm just looking to see if you were intending to iterate any further on this, or if I should make the updates to this myself after merging in your change.

Hi sorry! Yes will continue working on this. Going to add a test this week -- ideally one for each set. I'll also revert that unrelated change in the other function

**Comment by rjmholt** [comment]
I've found and fixed a bug, but when I try to open a PR from this master from my master it gives me a 404 (I'm pretty sure -- I tried it a few different ways).

If I could guarantee other branches it might be possible

**Comment by HowardWolosky** [comment]
>I've found and fixed a bug, but when I try to open a PR from this master from my master it gives me a 404 (I'm pretty sure -- I tried it a few different ways).
>
>If I could guarantee other branches it might be possible

@rjmholt - Apologies, I missed your previous comment until now.  I just tried to PR's using your current submitted code and it works fine for me.  Can you clarify the scenario that's giving you the `404`? I got a `422` if I tried to submit a PR for a branch that I already had an open PR for, but was unable to reproduce the `404` error you're referring to.

**Comment by HowardWolosky** [comment]
Hey @rjmholt -- just wanted to check-in with you on this one.

**Comment by rjmholt** [comment]
It looks like this for me:

```
 > New-GitHubPullRequest -OwnerName rjmholt -RepositoryName vscode-PowerShell -Title 'Banana' -Body 'Hi' -Head master -Base master
DONE - Operation took 2 second(s) [Creating pull request Banana in vscode-PowerShell]
Telemetry is currently enabled.  It can be disabled by calling "Set-GitHubConfiguration -DisableTelemetry". Refer to USAGE.md#telemetry for more information. Stop seeing this message in the future by calling "Set-GitHubConfiguration -SuppressTelemetryReminder".
{"message":"Validation Failed","errors":[{"resource":"PullRequest","code":"custom","message":"No commits between master and master"}],"documentation_url":"https://developer.github.com/v3/pulls/#create-a-pull-request"}
At /Users/rjmholt/Documents/Dev/Microsoft/PowerShellForGitHub/GitHubCore.ps1:314 char:17
+                 throw $remoteErrors[0].Exception
+                 ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+ CategoryInfo          : InvalidOperation: (Method: POST, Reque…n; charset=UTF-8
}:HttpRequestMessage) [Invoke-WebRequest], HttpResponseException
+ FullyQualifiedErrorId : WebCmdletWebResponseException,Microsoft.PowerShell.Commands.InvokeWebRequestCommand
```

Unless there's two branches that exist that we can guarantee commits between, I'm not sure what the best way to proceed with tests is (also just constantly creating PRs will drive the issue number up pretty heavily).

File: GitHubPullRequests.ps1 [hunk 5]
```diff
@@ -163,3 +163,229 @@ function Get-GitHubPullRequest
 
     return Invoke-GHRestMethodMultipleResult @params
 }
+
+function New-GitHubPullRequest
+{ 
+    <#
+    .SYNOPSIS
+        Create a new pull request in the specified repository.
+
```
Comment (HowardWolosky) [inline GitHubPullRequests.ps1]: nit: can you remove the trailing spaces from the empty lines?

**Comment by rjmholt** [comment]
> Hey -- I think at this point we abandon the idea of the Pester test with this one given the problems you're encountering.

Yeah -- sorry I wasn't able to get good tests working. It might be possible to mock `Invoke-GHRestMethod` to just verify the produced request body (at least that tests the logic of the wrapping cmdlet), but not sure how valuable that would be for the work.

> Thanks again, so much, for this contribution!

No worries, sorry it took so long

**Comment by HowardWolosky** [comment]
@rjmholt - I was reviewing this code today, and I can't find anywhere [in the documentation](https://developer.github.com/v3/pulls/#create-a-pull-request) that talks about creating an issue from a pull request:

https://github.com/microsoft/PowerShellForGitHub/blob/788465faec1b6d6331537aa87c2d94039682e373/GitHubPullRequests.ps1#L357-L361

Is this something that they've removed from the API since you added it?  Any idea where you got this idea from?  Thanks

**Pull request merged**